controllers/provider-aws/pkg/webhook/controlplane/mutator.go

```diff
@@ -156,3 +168,82 @@ func ensureVolumes(ps *corev1.PodSpec) {
 	ps.Volumes = controlplane.EnsureVolumeWithName(ps.Volumes, cloudProviderConfigVolume)
 	ps.Volumes = controlplane.EnsureVolumeWithName(ps.Volumes, cloudProviderSecretVolume)
 }
+
+func (m *mutator) mutateOperatingSystemConfig(osc *extensionsv1alpha1.OperatingSystemConfig) error {
+	// Mutate kubelet.service unit, if present
+	if u := controlplane.UnitWithName(osc.Spec.Units, "kubelet.service"); u != nil && u.Content != nil {
+		if err := m.ensureKubeletServiceUnitContent(u.Content); err != nil {
+			return err
+		}
+	}
+
+	// Mutate kubelet configuration file, if present
+	if f := controlplane.FileWithPath(osc.Spec.Files, "/var/lib/kubelet/config/kubelet"); f != nil && f.Content.Inline != nil {
+		if err := m.ensureKubeletConfigFileContent(f.Content.Inline); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *mutator) ensureKubeletServiceUnitContent(content *string) error {
+	var opts []*unit.UnitOption
+	var err error
+
+	// Deserialize unit options
+	if opts, err = m.unitSerializer.Deserialize(*content); err != nil {
+		return errors.Wrap(err, "could not deserialize kubelet.service unit content")
+	}
+
+	ensureKubeletServiceUnitOptions(opts)
+
+	// Serialize unit options
+	if *content, err = m.unitSerializer.Serialize(opts); err != nil {
+		return errors.Wrap(err, "could not serialize kubelet.service unit options")
+	}
+
+	return nil
+}
+
+func (m *mutator) ensureKubeletConfigFileContent(fci *extensionsv1alpha1.FileContentInline) error {
+	// Decode kubelet configuration from inline content
+	kubeletConfig := &kubeletconfigv1beta1.KubeletConfiguration{}
+	if err := m.kubeletConfigCodec.Decode(fci, kubeletConfig); err != nil {
+		return errors.Wrap(err, "could not decode kubelet configuration")
+	}
+
+	ensureKubeletConfiguration(kubeletConfig)
+
+	// Encode kubelet configuration into inline content
+	if err := m.kubeletConfigCodec.Encode(kubeletConfig, fci); err != nil {
+		return errors.Wrap(err, "could not encode kubelet configuration")
+	}
+
+	return nil
+}
+
+func ensureKubeletServiceUnitOptions(opts []*unit.UnitOption) {
+	if opt := controlplane.UnitOptionWithSectionAndName(opts, "Service", "ExecStart"); opt != nil {
+		command := controlplane.DeserializeCommandLine(opt.Value)
+		command = ensureKubeletCommandLineArgs(command)
+		opt.Value = controlplane.SerializeCommandLine(command, 1, " \\\n    ")
+	}
+}
+
+func ensureKubeletCommandLineArgs(command []string) []string {
+	command = controlplane.EnsureStringWithPrefix(command, "--cloud-provider=", "aws")
+	command = controlplane.EnsureNoStringWithPrefix(command, "--provider-id=")
```

Comment:
This is not needed I guess - Gardener won't add this flag.

Reply:
This is consistent with the behavior of this webhook with the 2 deployments. Here, as well as there, it makes sure that any potentially harmful flags / configuration bits are removed. It makes no assumptions what Gardener would do or would not do - wasn't this what we agreed?

Reply:
OK, I don't recall whether we talked about exactly this. It's definitely safer to have this logic. I was just worried that it would be hard to maintain with new Kubernetes versions, etc., because there are so many options that cannot be used in conjunction. We can keep it as it is now, and potentially revisit this in the future if we see that the complexity is increased. WDYT?

Reply:
Actually, I do think that without any assumptions about Gardener behavior the complexity of webhooks may explode. I think we should document what assumptions can be made in a README somewhere. The README should state what Gardener **will** do, what it **will not** do, and what it **may or may not** do (it could be safer not to make definite promises about some things). Such a README could play the role of a more explicit contract between Gardener and webhooks. This of course means that Gardener should try to abide by it in the future, breaking the contract only if absolutely necessary.

Reply:
Removed as initially suggested.